proxylib/kafka/parser.go

```diff
@@ -0,0 +1,286 @@
+		if want > have {
+			return want - have
+		}
+		return 1 // likely protocol error (invalid length)
+	}
+	return 4 - read // enough for the length field
+}
+
+func (r *Reader) Read(p []byte) (n int, err error) {
+	n = 0
+	l := len(p)
+	slices := len(r.buf)
+	for n < l && r.slice < slices {
+		nc := copy(p[n:], r.buf[r.slice][r.index:])
+		if nc == len(r.buf[r.slice][r.index:]) {
+			// next slice please
+			r.slice++
+			r.index = 0
+		} else {
+			// move ahead in the same slice
+			r.index += nc
+		}
+		n += nc
+	}
+	if n == 0 {
+		return 0, io.EOF
+	}
+	r.count += n
+	return n, nil
+}
+
+type KafkaParser struct {
+	connection *Connection
+}
+
+func (pf *KafkaParserFactory) Create(connection *Connection) Parser {
+	p := KafkaParser{connection: connection}
+	return &p
+}
+
+func (p *KafkaParser) OnData(reply, endStream bool, dataArray [][]byte) (OpType, int) {
+	reader := &Reader{buf: dataArray}
+
+	length := 0
+	for i := 0; i < len(dataArray); i++ {
+		length += len(dataArray[i])
+	}
+	if length == 0 {
+		return NOP, 0
+	}
+
+	if reply {
+		// Replies are parsed but always passed as-is.
+		// This allows the error responses to be injected on frame boundaries.
+		resp, err := kafka.ReadResponse(reader)
+		if err != nil {
+			if err == io.ErrUnexpectedEOF || err == io.EOF {
+				// Not enough data, ask for more and try again
+				return MORE, reader.Missing()
+			}
+			if flowdebug.Enabled() {
+				log.WithError(err).Warning("Unable to parse Kafka response; closing Kafka connection")
+			}
+			return ERROR, int(ERROR_INVALID_FRAME_TYPE)
+		}
+		p.connection.Log(cilium.EntryType_Response,
+			&cilium.LogEntry_GenericL7{
+				GenericL7: &cilium.L7LogEntry{
+					Proto: parserName,
+					Fields: map[string]string{
+						"CorrelationID": strconv.Itoa(int(resp.GetCorrelationID())),
+					},
+				},
+			})
+		return PASS, int(reader.count)
+	}
+
+	var req KafkaRequest
+	var err error
+	req.RequestMessage, err = kafka.ReadRequest(reader)
+	if err != nil {
+		if err == io.ErrUnexpectedEOF || err == io.EOF {
+			// Not enough data, ask for more and try again
+			return MORE, reader.Missing()
+		}
+		if flowdebug.Enabled() {
+			log.WithError(err).Warning("Unable to parse Kafka request; closing Kafka connection")
+		}
+		return ERROR, int(ERROR_INVALID_FRAME_TYPE)
+	}
+
+	topics := req.GetTopics()
+	// Maintain a map of all topics in the request.
+	// We should allow the request only if all topics are
+	// allowed by the list of rules.
+	req.topics = make(map[string]struct{}, len(topics))
+	for _, topic := range topics {
+		req.topics[topic] = struct{}{}
+	}
+	logEntry := &cilium.LogEntry_GenericL7{
+		GenericL7: &cilium.L7LogEntry{
+			Proto: parserName,
+			Fields: map[string]string{
+				"APIVersion":    strconv.Itoa(int(req.GetVersion())),
+				"APIKey":        apiKeyToString(req.GetAPIKey()),
+				"CorrelationID": strconv.Itoa(int(req.GetCorrelationID())),
+				"Topics":        strings.Join(topics, ","),
```

Comment:
these could be constants as well

Reply:
Will await for some more feedback on the general approach of using the "generic L7" format to internally carry Kafka log messages before doubling down on it :-)